### PR TITLE
Add tracking modes, update to pre10

### LINKED ISSUE #145: iOS: support `userTrackingMode`
The iOS SDK offers the `mapView.userTrackingMode` property to allow the map view to automatically center on and follow the user's coordinates. It looks like Android has support here via [`UserLocationTrackingMode`](https://github.com/mapbox/react-native-mapbox-gl/blob/3a7109b6ed0c3982dfc3b2cec100fa16f77e185e/index.android.js#L29), but the iOS implementation is missing.

[Here are the possible values](https://github.com/mapbox/mapbox-gl-native/blob/ddc4a8193e775d9cd9280ad7d9bb6975b4a323ee/include/mbgl/ios/MGLTypes.h#L17-L27) for `MGLMapView.userTrackingMode`:

``` objc
/** The mode used to track the user location on the map. */
typedef NS_ENUM(NSUInteger, MGLUserTrackingMode) {
    /** The map does not follow the user location. */
    MGLUserTrackingModeNone              = 0,
    /** The map follows the user location. */
    MGLUserTrackingModeFollow,
    /** The map follows the user location and rotates when the heading changes. */
    MGLUserTrackingModeFollowWithHeading,
    /** The map follows the user location and rotates when the course changes. */
    MGLUserTrackingModeFollowWithCourse,
};
```

/cc @bsudekum

### PR DESCRIPTION
**Pull request by bsudekum**
Adds user tracking modes, closes https://github.com/mapbox/react-native-mapbox-gl/issues/145 and https://github.com/mapbox/react-native-mapbox-gl/issues/142

@1ec5 @incanus I'm seeing an issue with `MGLUserTrackingModeFollowWithCourse` and `MGLUserTrackingModeFollowWithHeading`, namely that they don't work. `follow` seems to work fine but, even when I hard code `_map.userTrackingMode` to with course or heading, the map does not pan to the users location nor follow it with heading. Have you seen this before?

### CODE REVIEW

**Comment by 1ec5** [comment]
Are you testing in the Simulator? The Simulator doesn’t support heading simulation.

**Comment by bsudekum** [comment]
Ah yep

**Comment by bsudekum** [comment]
@1ec5 hmm on device, same issue, heading and course do not work... even when hard coded.

**Comment by incanus** [comment]
Another common thing that can turn this off is any sort of programmatic or gestural movement of the map. The best way to figure out how it’s happening is to set a breakpoint in the setter for the mode and then see what triggers it to turn off, or if it’s truly just not being set in the first place for some reason.

**Comment by bsudekum** [comment]
Although not ideal, I added a timeout to setting the tracking mode. This seems to do the trick for now, I'm going to revisit this later.

File: ios/RCTMapboxGL/RCTMapboxGLManager.m [hunk 1]
```diff
@@ -46,16 +46,24 @@ - (NSArray *)customDirectEventTypes
...
+             @"mapStyles": @{
                  @"light": [[MGLStyle lightStyleURL] absoluteString],
                  @"dark": [[MGLStyle darkStyleURL] absoluteString],
                  @"streets": [[MGLStyle streetsStyleURL] absoluteString],
                  @"emerald": [[MGLStyle emeraldStyleURL] absoluteString],
                  @"satellite": [[MGLStyle satelliteStyleURL] absoluteString],
                  @"hybrid": [[MGLStyle hybridStyleURL] absoluteString],
+            },
+             @"userTrackingMode": @{
+                     @"none": [NSNumber numberWithUnsignedInt:MGLUserTrackingModeNone],
```
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGLManager.m:58]: These can be written as simply `@(MGLUserTrackingModeNone)`.

File: ios/RCTMapboxGL/RCTMapboxGL.m [hunk 2]
```diff
@@ -217,6 +219,16 @@ - (void)setStyleURL:(NSURL *)styleURL
     [self updateMap];
 }
 
+- (void)setUserTrackingMode:(int)userTrackingMode
+{
+    if (userTrackingMode > 3 || userTrackingMode < 0) {
+        _userTrackingMode = 0;
+    } else {
+        _userTrackingMode = userTrackingMode;
```
Comment (1ec5) [inline ios/RCTMapboxGL/RCTMapboxGL.m:227]: Should issue an error and return in this case.

**Comment by 1ec5** [comment]
I’m guessing this is the reason heading or course tracking doesn’t activate when you attempt to enable them synchronously: as soon as you try to switch to heading or course tracking, [this line](https://github.com/mapbox/mapbox-gl-native/blob/b3dbdb389ff712ace00857e3e72894f0bb3afe3a/platform/ios/MGLMapView.mm#L2660) resets the mode to no tracking if the map view has yet to obtain a valid user location. Effectively, this means that you need to have entered follow mode at least once before entering heading or course tracking mode.

iosapp [shows the user dot _after_ setting the tracking mode](https://github.com/mapbox/mapbox-gl-native/blob/8a0bce780847546c161670420e4d0f1e3a6cbdca/ios/app/MBXViewController.mm#L115). By a fit of Objective-C luck, the map view’s `userLocation` is therefore `nil`, so `userLocation.coordinate` is (0, 0), which is a valid coordinate.

We ported this check from the legacy iOS SDK, where it was introduced in mapbox/mapbox-ios-sdk-legacy@f781914dcc2727e9ab974f2f257c4ee02d38b86d. @incanus, is this behavior still relevant?

**Pull request merged**